{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ce412e60-d13f-47cb-b938-632a2f36d43d",
   "metadata": {},
   "source": [
    "# eQTL mapping example for n=94"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3bce7902-ba74-4a07-ab5e-9c57345baff3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "from jax.config import config\n",
    "\n",
    "from jaxqtl.families.distribution import Poisson, Gaussian\n",
    "from jaxqtl.io.geno import PlinkReader\n",
    "from jaxqtl.io.pheno import PheBedReader\n",
    "from jaxqtl.io.readfile import create_readydata\n",
    "from jaxqtl.io.covar import covar_reader\n",
    "from jaxqtl.map import map_cis, map_cis_nominal\n",
    "from jaxqtl.infer.permutation import DirectPerm, Permutation\n",
    "from jaxqtl.log import get_log\n",
    "\n",
    "config.update(\"jax_enable_x64\", True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7bf395b-7651-49bc-afd9-47eed4f24e5f",
   "metadata": {},
   "source": [
    "### Read data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "05d9f503-b2d9-4b15-8ebf-b93ac6b16245",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "geno_path = \"../example/data/chr22.n94.bed\"\n",
    "covar_path = \"../example/data/donor_features.n94.tsv\"\n",
    "pheno_path = \"../example/data/CD14_positive_monocyte.bed.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5a9da72e-d647-444c-8ebf-c34b95bc8e64",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/zixuanzhang/Dropbox/PhD_Projects/Nick/jaxqtl/src/jaxqtl/io/readfile.py:92: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  pheno.drop([\"chr\", \"start\", \"end\"], axis=1, inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# raw genotype data and impute for genotype data\n",
    "geno_reader = PlinkReader()\n",
    "geno, bim, sample_info = geno_reader(geno_path)\n",
    "\n",
    "covar = covar_reader(covar_path)\n",
    "\n",
    "pheno_reader = PheBedReader()\n",
    "pheno = pheno_reader(pheno_path)\n",
    "\n",
    "# run Mapping #\n",
    "dat = create_readydata(\n",
    "    geno,\n",
    "    bim,\n",
    "    pheno,\n",
    "    covar,\n",
    "    autosomal_only = True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e28dbbc4-c15d-4ec3-bf66-7a1e5fa278f6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# add first 10 PC of expression profile for this cell type\n",
    "dat.add_covar_pheno_PC(k=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1d438c8-018d-416c-831b-66e1991cf276",
   "metadata": {},
   "source": [
    "### Cis Mapping (Report lead eQTL for each gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ffc480f2-55a4-468a-bb80-1b22685e9c3a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# cis-mapping for chr22 alone\n",
    "dat.filter_geno(0., \"22\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c8c9e142-534a-4ed5-80de-dcf495d9200b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(94, 143083)\n",
      "(143083, 7)\n",
      "(6990, 4)\n",
      "<StringArray>\n",
      "['22']\n",
      "Length: 1, dtype: string\n",
      "(94, 14)\n"
     ]
    }
   ],
   "source": [
    "print(dat.geno.shape) # total 143083 variants\n",
    "print(dat.bim.shape)\n",
    "print(dat.pheno_meta.gene_map.shape) # 200 genes\n",
    "print(dat.bim.chrom.unique())\n",
    "print(dat.covar.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e5338d9-fd31-4698-9478-fc6356c194ab",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Default is fitting poisson model and report nominal and adjusted p value using beta distribution method\n",
    "# for unit testing, run cis-mapping for the first two genes\n",
    "%time mapcis_df = map_cis(dat, family=Poisson(), test_break_n=2, direct_perm=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c6513713-557c-417b-b9df-ff213f345544",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>num_var</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>tss_distance</th>\n",
       "      <th>beta_shape1</th>\n",
       "      <th>beta_shape2</th>\n",
       "      <th>beta_converged</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>af</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>pval_perm</th>\n",
       "      <th>pval_beta</th>\n",
       "      <th>qval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>22</td>\n",
       "      <td>2592</td>\n",
       "      <td>22:17691970</td>\n",
       "      <td>126126</td>\n",
       "      <td>1.050303</td>\n",
       "      <td>31.130028</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.989362</td>\n",
       "      <td>0.072695</td>\n",
       "      <td>-1.039481</td>\n",
       "      <td>0.579182</td>\n",
       "      <td>0.904762</td>\n",
       "      <td>0.896194</td>\n",
       "      <td>0.896194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000069998</td>\n",
       "      <td>22</td>\n",
       "      <td>2862</td>\n",
       "      <td>22:17826030</td>\n",
       "      <td>179853</td>\n",
       "      <td>9.915831</td>\n",
       "      <td>57.407599</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.984043</td>\n",
       "      <td>0.118103</td>\n",
       "      <td>-1.225258</td>\n",
       "      <td>0.784021</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>0.263327</td>\n",
       "      <td>0.789982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000093072</td>\n",
       "      <td>22</td>\n",
       "      <td>3015</td>\n",
       "      <td>22:18123765</td>\n",
       "      <td>420886</td>\n",
       "      <td>2.333648</td>\n",
       "      <td>78.253388</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.994681</td>\n",
       "      <td>0.033450</td>\n",
       "      <td>-1.368323</td>\n",
       "      <td>0.643419</td>\n",
       "      <td>0.619048</td>\n",
       "      <td>0.668867</td>\n",
       "      <td>0.896194</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      phenotype_id chrom  num_var   variant_id  tss_distance  beta_shape1   \n",
       "0  ENSG00000177663    22     2592  22:17691970        126126     1.050303  \\\n",
       "1  ENSG00000069998    22     2862  22:17826030        179853     9.915831   \n",
       "2  ENSG00000093072    22     3015  22:18123765        420886     2.333648   \n",
       "\n",
       "   beta_shape2  beta_converged  ma_samples  ma_count        af  pval_nominal   \n",
       "0    31.130028             1.0           2       2.0  0.989362      0.072695  \\\n",
       "1    57.407599             1.0           3       3.0  0.984043      0.118103   \n",
       "2    78.253388             1.0           1       1.0  0.994681      0.033450   \n",
       "\n",
       "      slope  slope_se  pval_perm  pval_beta      qval  \n",
       "0 -1.039481  0.579182   0.904762   0.896194  0.896194  \n",
       "1 -1.225258  0.784021   0.380952   0.263327  0.789982  \n",
       "2 -1.368323  0.643419   0.619048   0.668867  0.896194  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# shifted transform count, y / s + 1\n",
    "mapcis_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "13f7dbe9-ae8a-446f-b298-bb99d82b002c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>num_var</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>tss_distance</th>\n",
       "      <th>beta_shape1</th>\n",
       "      <th>beta_shape2</th>\n",
       "      <th>beta_converged</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>af</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>pval_perm</th>\n",
       "      <th>pval_beta</th>\n",
       "      <th>qval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>22</td>\n",
       "      <td>2592</td>\n",
       "      <td>22:17691970</td>\n",
       "      <td>126126</td>\n",
       "      <td>1.050303</td>\n",
       "      <td>31.130028</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.989362</td>\n",
       "      <td>0.072695</td>\n",
       "      <td>-1.039481</td>\n",
       "      <td>0.579182</td>\n",
       "      <td>0.904762</td>\n",
       "      <td>0.896194</td>\n",
       "      <td>0.896194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000069998</td>\n",
       "      <td>22</td>\n",
       "      <td>2862</td>\n",
       "      <td>22:17826030</td>\n",
       "      <td>179853</td>\n",
       "      <td>9.915831</td>\n",
       "      <td>57.407599</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.984043</td>\n",
       "      <td>0.118103</td>\n",
       "      <td>-1.225258</td>\n",
       "      <td>0.784021</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>0.263327</td>\n",
       "      <td>0.789982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000093072</td>\n",
       "      <td>22</td>\n",
       "      <td>3015</td>\n",
       "      <td>22:18123765</td>\n",
       "      <td>420886</td>\n",
       "      <td>2.333648</td>\n",
       "      <td>78.253388</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.994681</td>\n",
       "      <td>0.033450</td>\n",
       "      <td>-1.368323</td>\n",
       "      <td>0.643419</td>\n",
       "      <td>0.619048</td>\n",
       "      <td>0.668867</td>\n",
       "      <td>0.896194</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      phenotype_id chrom  num_var   variant_id  tss_distance  beta_shape1   \n",
       "0  ENSG00000177663    22     2592  22:17691970        126126     1.050303  \\\n",
       "1  ENSG00000069998    22     2862  22:17826030        179853     9.915831   \n",
       "2  ENSG00000093072    22     3015  22:18123765        420886     2.333648   \n",
       "\n",
       "   beta_shape2  beta_converged  ma_samples  ma_count        af  pval_nominal   \n",
       "0    31.130028             1.0           2       2.0  0.989362      0.072695  \\\n",
       "1    57.407599             1.0           3       3.0  0.984043      0.118103   \n",
       "2    78.253388             1.0           1       1.0  0.994681      0.033450   \n",
       "\n",
       "      slope  slope_se  pval_perm  pval_beta      qval  \n",
       "0 -1.039481  0.579182   0.904762   0.896194  0.896194  \n",
       "1 -1.225258  0.784021   0.380952   0.263327  0.789982  \n",
       "2 -1.368323  0.643419   0.619048   0.668867  0.896194  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# shifted transform count, y / s + 1\n",
    "mapcis_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da0c31d5-9e51-4d03-891e-3cef761958a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapcis_df_Gaussian = map_cis(dat, family=Gaussian())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "833d361c-5d65-4a5c-b2ee-ec90bf39bd0c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from jaxqtl.post.qvalue import add_qvalues\n",
    "import numpy as np\n",
    "from jaxqtl.log import get_log\n",
    "log = get_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5ac73c86-a0c5-495c-a2ac-f59f19f8f915",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>num_var</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>tss_distance</th>\n",
       "      <th>beta_shape1</th>\n",
       "      <th>beta_shape2</th>\n",
       "      <th>beta_converged</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>af</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>pval_perm</th>\n",
       "      <th>pval_beta</th>\n",
       "      <th>qval</th>\n",
       "      <th>pval_nominal_threshold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>22</td>\n",
       "      <td>2592</td>\n",
       "      <td>22:17691970</td>\n",
       "      <td>126126</td>\n",
       "      <td>2.681706</td>\n",
       "      <td>1.063671</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.989362</td>\n",
       "      <td>0.072695</td>\n",
       "      <td>-1.039481</td>\n",
       "      <td>0.579182</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.000983</td>\n",
       "      <td>0.001404</td>\n",
       "      <td>0.110781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000069998</td>\n",
       "      <td>22</td>\n",
       "      <td>2862</td>\n",
       "      <td>22:17826030</td>\n",
       "      <td>179853</td>\n",
       "      <td>3.566182</td>\n",
       "      <td>1.179464</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.984043</td>\n",
       "      <td>0.118103</td>\n",
       "      <td>-1.225258</td>\n",
       "      <td>0.784021</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.000675</td>\n",
       "      <td>0.001126</td>\n",
       "      <td>0.180498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000093072</td>\n",
       "      <td>22</td>\n",
       "      <td>3015</td>\n",
       "      <td>22:18123765</td>\n",
       "      <td>420886</td>\n",
       "      <td>2.316630</td>\n",
       "      <td>1.154663</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.994681</td>\n",
       "      <td>0.033450</td>\n",
       "      <td>-1.368323</td>\n",
       "      <td>0.643419</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.000481</td>\n",
       "      <td>0.000963</td>\n",
       "      <td>0.074206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000131100</td>\n",
       "      <td>22</td>\n",
       "      <td>3589</td>\n",
       "      <td>22:18525678</td>\n",
       "      <td>414094</td>\n",
       "      <td>2.370120</td>\n",
       "      <td>0.969281</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.994681</td>\n",
       "      <td>0.007647</td>\n",
       "      <td>-1.806344</td>\n",
       "      <td>0.677224</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.088437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000099968</td>\n",
       "      <td>22</td>\n",
       "      <td>3590</td>\n",
       "      <td>22:17641685</td>\n",
       "      <td>-469936</td>\n",
       "      <td>3.313077</td>\n",
       "      <td>1.147476</td>\n",
       "      <td>1.0</td>\n",
       "      <td>26</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.861702</td>\n",
       "      <td>0.152060</td>\n",
       "      <td>-0.657656</td>\n",
       "      <td>0.459163</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>0.002505</td>\n",
       "      <td>0.002783</td>\n",
       "      <td>0.161186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000015475</td>\n",
       "      <td>22</td>\n",
       "      <td>3191</td>\n",
       "      <td>22:18043325</td>\n",
       "      <td>-214211</td>\n",
       "      <td>3.347359</td>\n",
       "      <td>1.170385</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.994681</td>\n",
       "      <td>0.128228</td>\n",
       "      <td>-1.018026</td>\n",
       "      <td>0.669258</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.001384</td>\n",
       "      <td>0.001730</td>\n",
       "      <td>0.162388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000269220</td>\n",
       "      <td>22</td>\n",
       "      <td>3190</td>\n",
       "      <td>22:18062511</td>\n",
       "      <td>-197577</td>\n",
       "      <td>3.031552</td>\n",
       "      <td>1.103206</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.973404</td>\n",
       "      <td>0.064286</td>\n",
       "      <td>-1.498019</td>\n",
       "      <td>0.809656</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.000291</td>\n",
       "      <td>0.000868</td>\n",
       "      <td>0.139586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000070413</td>\n",
       "      <td>22</td>\n",
       "      <td>2647</td>\n",
       "      <td>22:19277323</td>\n",
       "      <td>167356</td>\n",
       "      <td>3.724353</td>\n",
       "      <td>1.287164</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.984043</td>\n",
       "      <td>0.041943</td>\n",
       "      <td>-1.752298</td>\n",
       "      <td>0.861467</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.185124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000100075</td>\n",
       "      <td>22</td>\n",
       "      <td>2782</td>\n",
       "      <td>22:19428145</td>\n",
       "      <td>261802</td>\n",
       "      <td>4.668000</td>\n",
       "      <td>1.267320</td>\n",
       "      <td>1.0</td>\n",
       "      <td>41</td>\n",
       "      <td>47.0</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.260392</td>\n",
       "      <td>-0.611879</td>\n",
       "      <td>0.543668</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.003035</td>\n",
       "      <td>0.003035</td>\n",
       "      <td>0.260392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000185608</td>\n",
       "      <td>22</td>\n",
       "      <td>3437</td>\n",
       "      <td>22:19660536</td>\n",
       "      <td>241111</td>\n",
       "      <td>3.190736</td>\n",
       "      <td>1.098021</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.994681</td>\n",
       "      <td>0.078031</td>\n",
       "      <td>-1.597082</td>\n",
       "      <td>0.906285</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.000347</td>\n",
       "      <td>0.000868</td>\n",
       "      <td>0.154259</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      phenotype_id chrom  num_var   variant_id  tss_distance  beta_shape1   \n",
       "0  ENSG00000177663    22     2592  22:17691970        126126     2.681706  \\\n",
       "1  ENSG00000069998    22     2862  22:17826030        179853     3.566182   \n",
       "2  ENSG00000093072    22     3015  22:18123765        420886     2.316630   \n",
       "3  ENSG00000131100    22     3589  22:18525678        414094     2.370120   \n",
       "4  ENSG00000099968    22     3590  22:17641685       -469936     3.313077   \n",
       "5  ENSG00000015475    22     3191  22:18043325       -214211     3.347359   \n",
       "6  ENSG00000269220    22     3190  22:18062511       -197577     3.031552   \n",
       "7  ENSG00000070413    22     2647  22:19277323        167356     3.724353   \n",
       "8  ENSG00000100075    22     2782  22:19428145        261802     4.668000   \n",
       "9  ENSG00000185608    22     3437  22:19660536        241111     3.190736   \n",
       "\n",
       "   beta_shape2  beta_converged  ma_samples  ma_count        af  pval_nominal   \n",
       "0     1.063671             1.0           2       2.0  0.989362      0.072695  \\\n",
       "1     1.179464             1.0           3       3.0  0.984043      0.118103   \n",
       "2     1.154663             1.0           1       1.0  0.994681      0.033450   \n",
       "3     0.969281             1.0           1       1.0  0.994681      0.007647   \n",
       "4     1.147476             1.0          26      26.0  0.861702      0.152060   \n",
       "5     1.170385             1.0           1       1.0  0.994681      0.128228   \n",
       "6     1.103206             1.0           5       5.0  0.973404      0.064286   \n",
       "7     1.287164             1.0           3       3.0  0.984043      0.041943   \n",
       "8     1.267320             1.0          41      47.0  0.250000      0.260392   \n",
       "9     1.098021             1.0           1       1.0  0.994681      0.078031   \n",
       "\n",
       "      slope  slope_se  pval_perm  pval_beta      qval  pval_nominal_threshold  \n",
       "0 -1.039481  0.579182   0.047619   0.000983  0.001404                0.110781  \n",
       "1 -1.225258  0.784021   0.047619   0.000675  0.001126                0.180498  \n",
       "2 -1.368323  0.643419   0.047619   0.000481  0.000963                0.074206  \n",
       "3 -1.806344  0.677224   0.047619   0.000009  0.000062                0.088437  \n",
       "4 -0.657656  0.459163   0.095238   0.002505  0.002783                0.161186  \n",
       "5 -1.018026  0.669258   0.047619   0.001384  0.001730                0.162388  \n",
       "6 -1.498019  0.809656   0.047619   0.000291  0.000868                0.139586  \n",
       "7 -1.752298  0.861467   0.047619   0.000012  0.000062                0.185124  \n",
       "8 -0.611879  0.543668   0.047619   0.003035  0.003035                0.260392  \n",
       "9 -1.597082  0.906285   0.047619   0.000347  0.000868                0.154259  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pi0 = None\n",
    "qvalue_lambda = None\n",
    "add_qvalues(mapcis_df, log, 0.05, pi0, qvalue_lambda)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee388dac-6e4f-4b5a-85f6-377740438cfe",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Cis Mapping (Report all cis association statistics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4e9fc02e-28ff-4c6a-89c3-9a9792f57049",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/zixuanzhang/Dropbox/PhD_Projects/Nick/jaxqtl/src/jaxqtl/map.py:398: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  outdf[\"af\"][start_row:end_row] = af[idx]\n",
      "/Users/zixuanzhang/Dropbox/PhD_Projects/Nick/jaxqtl/src/jaxqtl/map.py:399: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  outdf[\"ma_samples\"][start_row:end_row] = ma_samples[idx]\n",
      "/Users/zixuanzhang/Dropbox/PhD_Projects/Nick/jaxqtl/src/jaxqtl/map.py:400: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  outdf[\"ma_count\"][start_row:end_row] = ma_count[idx]\n",
      "/Users/zixuanzhang/Dropbox/PhD_Projects/Nick/jaxqtl/src/jaxqtl/map.py:401: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  outdf[\"pval_nominal\"][start_row:end_row] = nominal_p[idx]\n",
      "/Users/zixuanzhang/Dropbox/PhD_Projects/Nick/jaxqtl/src/jaxqtl/map.py:402: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  outdf[\"slope\"][start_row:end_row] = slope[idx]\n",
      "/Users/zixuanzhang/Dropbox/PhD_Projects/Nick/jaxqtl/src/jaxqtl/map.py:403: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  outdf[\"slope_se\"][start_row:end_row] = slope_se[idx]\n",
      "/Users/zixuanzhang/Dropbox/PhD_Projects/Nick/jaxqtl/src/jaxqtl/map.py:404: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  outdf[\"converged\"][start_row:end_row] = converged[idx]\n"
     ]
    }
   ],
   "source": [
    "out_path = \"../example/result/dat_n94\"\n",
    "log=get_log()\n",
    "\n",
    "map_cis_nominal(dat, family=Poisson(), out_path=out_path, log=log, test_break=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b017da97-b0a8-4a10-9d06-06675f0fce0b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>pos</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>tss_distance</th>\n",
       "      <th>af</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>converged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>22:17066020</td>\n",
       "      <td>17066020</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>-499824</td>\n",
       "      <td>0.984043</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.840664</td>\n",
       "      <td>0.265719</td>\n",
       "      <td>1.321694</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>22:17066700</td>\n",
       "      <td>17066700</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>-499144</td>\n",
       "      <td>0.760638</td>\n",
       "      <td>40.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.669325</td>\n",
       "      <td>-0.115669</td>\n",
       "      <td>0.270840</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>22:17067504</td>\n",
       "      <td>17067504</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>-498340</td>\n",
       "      <td>0.313830</td>\n",
       "      <td>51.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>0.837865</td>\n",
       "      <td>0.052088</td>\n",
       "      <td>0.254553</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>22:17068400</td>\n",
       "      <td>17068400</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>-497444</td>\n",
       "      <td>0.117021</td>\n",
       "      <td>21.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.740131</td>\n",
       "      <td>0.114253</td>\n",
       "      <td>0.344468</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>22:17069064</td>\n",
       "      <td>17069064</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>-496780</td>\n",
       "      <td>0.111702</td>\n",
       "      <td>21.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.660578</td>\n",
       "      <td>0.171499</td>\n",
       "      <td>0.390555</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom          snp       pos     phenotype_id  tss_distance        af   \n",
       "0    22  22:17066020  17066020  ENSG00000177663       -499824  0.984043  \\\n",
       "1    22  22:17066700  17066700  ENSG00000177663       -499144  0.760638   \n",
       "2    22  22:17067504  17067504  ENSG00000177663       -498340  0.313830   \n",
       "3    22  22:17068400  17068400  ENSG00000177663       -497444  0.117021   \n",
       "4    22  22:17069064  17069064  ENSG00000177663       -496780  0.111702   \n",
       "\n",
       "   ma_samples  ma_count  pval_nominal     slope  slope_se  converged  \n",
       "0         3.0       3.0      0.840664  0.265719  1.321694       True  \n",
       "1        40.0      45.0      0.669325 -0.115669  0.270840       True  \n",
       "2        51.0      59.0      0.837865  0.052088  0.254553       True  \n",
       "3        21.0      22.0      0.740131  0.114253  0.344468       True  \n",
       "4        21.0      21.0      0.660578  0.171499  0.390555       True  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# shifted transform result\n",
    "pairs_df = pd.read_parquet(out_path+f'.cis_qtl_pairs.22.parquet')\n",
    "pairs_df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
